### PR TITLE
chore: introduced check for emitting event correctly

### PR DESCRIPTION
**Pull request by iamsahu**
## Description
This pull request introduces important fixes and improvements to the `WLSETHV1` contract and its tests, primarily focusing on correct event emission and allowance handling during transfers, especially in the context of rebasing and edge cases where the underlying value does not map to shares.

**Core contract fixes and improvements:**

* **Transfer and Allowance Handling:**
  - The `transfer` and `transferFrom` functions now properly return `true` on success, and `transferFrom` is protected against reentrancy. Allowance checks and spending are corrected to ensure atomicity and proper error handling.
  - The internal `_spendAllowance` method now requires the current allowance to be passed in, reducing redundant storage reads and improving clarity.

* **Event Emission and Edge Case Handling:**
  - The `_transfer` function now emits a `Transfer` event with the correct underlying value. If the value to shares conversion yields zero (i.e., the transfer amount is too small), a `Transfer` event with value `0` is emitted, and balances remain unchanged.

**Testing enhancements:**

* **Comprehensive Event Tests:**
  - Added a suite of tests to ensure that `Transfer` events always emit the correct underlying value after rebasing, for `mint`, `burn`, `transfer`, and `transferFrom` operations.
  - Added tests to verify that when the transfer amount is too small to convert to shares, a `Transfer` event with value `0` is emitted and no balances or allowances are changed.

These changes improve the correctness, safety, and transparency of the contract, especially in scenarios involving rebasing and small-value transfers.
## Notice

- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/liquid-collective/liquid-collective-protocol/pulls) for the same update/change?
- [ ] Have you assigned this PR to yourself?
- [ ] Have you added at least 1 reviewer?
- [ ] Have you updated the official documentation?
- [ ] Have you added sufficient documentation in your code?
- [ ] Have you added relevant tests to the official test suite?

## Pull Request Type

- [ ] 💫 New Feature (Breaking Change)
- [ ] 💫 New Feature (Non-breaking Change)
- [ ] 🛠️ Bug fix (Non-breaking Change: Fixes an issue)
- [ ] 🕹️ Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)

## Testing

- [ ] Have you tested this code with the official test suite?
- [ ] Have you tested this code manually?

## Summary by CodeRabbit

* **Bug Fixes**
  * Transfer events now report underlying token amounts after rebases; when conversion yields zero shares a zero-value Transfer is emitted and balances/allowances remain unchanged.
  * Allowances are only consumed when a transfer actually moves non-zero shares; no-op transfers no longer modify state.

* **Tests**
  * Added tests verifying Transfer events for mint, burn, transfer, and transferFrom reflect underlying amounts and cover rounding-to-zero edge cases.

### CODE REVIEW

File: contracts/src/WLSETH.1.sol [hunk 1]
```diff
@@ -211,10 +211,14 @@ contract WLSETHV1 is IWLSETHV1, Initializable, ReentrancyGuardUpgradeable {
...
         }
         uint256 valueToShares = river.sharesFromUnderlyingBalance(_value);
-        BalanceOf.set(_from, BalanceOf.get(_from) - valueToShares);
-        BalanceOf.set(_to, BalanceOf.get(_to) + valueToShares);
+        if (valueToShares > 0) {
+            BalanceOf.set(_from, BalanceOf.get(_from) - valueToShares);
+            BalanceOf.set(_to, BalanceOf.get(_to) + valueToShares);
 
-        emit Transfer(_from, _to, _value);
+            emit Transfer(_from, _to, _value);
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/src/WLSETH.1.sol:227]: Since, the `_value` is in ETH emitting `_value` is the right thing to do.

File: contracts/src/WLSETH.1.sol [hunk 2]
```diff
@@ -211,10 +211,14 @@ contract WLSETHV1 is IWLSETHV1, Initializable, ReentrancyGuardUpgradeable {
...
+        } else {
+            emit Transfer(_from, _to, 0);
+        }
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/src/WLSETH.1.sol:232]: This has been fixed.

File: contracts/src/WLSETH.1.sol [hunk 3]
```diff
@@ -211,11 +215,15 @@ contract WLSETHV1 is IWLSETHV1, Initializable, ReentrancyGuardUpgradeable {
...
+        if (valueToShares > 0) {
+            BalanceOf.set(_from, BalanceOf.get(_from) - valueToShares);
+            BalanceOf.set(_to, BalanceOf.get(_to) + valueToShares);
+
+            emit Transfer(_from, _to, _value);
+            return true;
+        } else {
+            emit Transfer(_from, _to, 0);
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/src/WLSETH.1.sol:232]: `_value` is the underlying balance that's why the current implementation is correct.

File: contracts/src/WLSETH.1.sol [hunk 4]
```diff
@@ -211,11 +215,15 @@ contract WLSETHV1 is IWLSETHV1, Initializable, ReentrancyGuardUpgradeable {
...
-
-        return true;
+        if (valueToShares > 0) {
+            BalanceOf.set(_from, BalanceOf.get(_from) - valueToShares);
+            BalanceOf.set(_to, BalanceOf.get(_to) + valueToShares);
+
+            emit Transfer(_from, _to, _value);
+            return true;
+        } else {
+            emit Transfer(_from, _to, 0);
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/src/WLSETH.1.sol:230]: We don't want to do this as we always want to emit the value in ETH. In case of the functions `transfer` & `transferFrom` the `_value` is passed in ETH so emitting that is consistent with what happens in mint & burn.

File: contracts/src/WLSETH.1.sol [hunk 5]
```diff
@@ -101,8 +102,11 @@ contract WLSETHV1 is IWLSETHV1, Initializable, ReentrancyGuardUpgradeable {
         if (_to == address(0)) {
             revert UnauthorizedTransfer(_from, address(0));
         }
-        _spendAllowance(_from, _value);
-        return _transfer(_from, _to, _value);
+        bool movedShares = _transfer(_from, _to, _value);
+        if (movedShares) {
+            _spendAllowance(_from, _value);
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/src/WLSETH.1.sol]: Since `_value` is in ETH the implementation is correct.

File: contracts/test/WLSETH.1.t.sol [hunk 6]
```diff
@@ -849,6 +849,170 @@ contract WLSETHV1Tests is WLSETHV1TestBase {
...
+        // Transfer 1 wei via transferFrom, converts to 0 shares
+        vm.startPrank(_approved);
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(_from, _recipient, 0);
+        wlseth.transferFrom(_from, _recipient, 1);
+        vm.stopPrank();
+
+        // Balances unchanged since 0 shares transferred
+        assert(wlseth.balanceOf(_recipient) == 0);
+    }
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/test/WLSETH.1.t.sol:1035]: Fixed in test

File: contracts/test/WLSETH.1.t.sol [hunk 7]
```diff
@@ -849,6 +849,173 @@ contract WLSETHV1Tests is WLSETHV1TestBase {
...
+        // User has 1000 ether balance now, try to transfer 1 wei
+        // This converts to 0 shares, so Transfer event emits 0
+        vm.startPrank(_guy);
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(_guy, _recipient, 0);
+        wlseth.transfer(_recipient, 1);
+        vm.stopPrank();
+
+        // Balances unchanged since 0 shares transferred
+        assert(wlseth.balanceOf(_recipient) == 0);
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/test/WLSETH.1.t.sol:995]: Fixed

File: contracts/src/WLSETH.1.sol [hunk 8]
```diff
@@ -101,8 +102,11 @@ contract WLSETHV1 is IWLSETHV1, Initializable, ReentrancyGuardUpgradeable {
...
+        }
+        return true;
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/src/WLSETH.1.sol:117]: Fixed.

**Review by mischat — DISMISSED** [review]
LGTM

File: contracts/test/WLSETH.1.t.sol [hunk 9]
```diff
@@ -849,6 +849,190 @@ contract WLSETHV1Tests is WLSETHV1TestBase {
...
+
+        vm.prank(_from);
+        wlseth.approve(_approved, 1);
+
+        // Transfer 1 wei via transferFrom, converts to 0 shares
+        vm.startPrank(_approved);
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(_from, _recipient, 0);
+        wlseth.transferFrom(_from, _recipient, 1);
+        vm.stopPrank();
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/test/WLSETH.1.t.sol:1027]: Added the test

File: contracts/src/WLSETH.1.sol [hunk 10]
```diff
@@ -168,13 +180,10 @@ contract WLSETHV1 is IWLSETHV1, Initializable, ReentrancyGuardUpgradeable {
...
-        if (currentAllowance < _value) {
-            revert AllowanceTooLow(_from, msg.sender, currentAllowance, _value);
-        }
-        if (currentAllowance != type(uint256).max) {
-            _approve(_from, msg.sender, currentAllowance - _value);
+    /// @param _currentAllowance Pre-fetched current allowance (must be >= _value)
+    function _spendAllowance(address _from, uint256 _value, uint256 _currentAllowance) internal {
+        if (_currentAllowance != type(uint256).max) {
+            _approve(_from, msg.sender, _currentAllowance - _value);
         }
```
[the remark this answers is not in the document]
Comment (iamsahu) [reply contracts/src/WLSETH.1.sol:187]: Will not add this.

**Review by mischat-galaxy — APPROVED** [review]
i am happy with this now, thanks!

**Review by mischat — APPROVED** [review]
🦈
